source/manual/intro-to-docker.html.md

```diff
@@ -0,0 +1,331 @@
+```
+
+*   Build our new image:
+
+```shell
+$mac docker build -t content-publisher
+```
+
+This creates an image based on the Dockerfile we just added.
+
+`-t` tags, or names it, as `content-publisher`.
+
+*   Now we can start a container with our new image:
+
+```shell
+$mac docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle content-publisher bash
+```
+
+Each time you change the Dockerfile, you need to remember to rebuild the docker image.
+
+If we run the tests again, this time we see another problem about no database. That's because we haven't got PostgreSQL installed.
+
+#### Use PostgreSQL image/container
+
+Docker containers are quick to start and it's possible for them to talk to each other, so let's set up a separate container for postgres.
+
+*  In another terminal, if we run the following to start a new container based on the postgres image:
+
+```shell
+$mac docker run -it --rm postgres
+```
+
+* Now to see what containers you have, run:
+
+```shell
+$mac docker ps
+```
+
+And you should see (ignore any other containers you've got running):
+
+```shell
+$mac
+CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
+a7b87be3e1bc        postgres            "docker-entrypoint.s…"   21 minutes ago      Up 21 minutes       5432/tcp            jovial_matsumoto
+```
+
+* Find the ID for the postgres container, and with it run:
+
+```shell
+$mac docker inspect [container ID for postgres]
+```
+
+Grab the IP address from the output. We will use that IP address to point the content-publisher container to the one running postgres. Rails will use this to run its tests against.
+
+
+*   Set these environment variables:
+
+```shell
+$dev export DATABASE_URL=postgres://postgres@172.17.0.3/content-publisher-dev
+$dev export TEST_DATABASE_URL=postgres://postgres@172.17.0.3/content-publisher-test
+```
+
+> NOTE: The IP address will change randomly so you may need to update the IP address above each time you restart the container! See [Fixing IP with Docker Network]
+
+*   Run the tests in content-publisher container:
+
+```shell
+$dev bundle exec rake db:setup
+$dev bundle exec rake
+```
+
+We have a database, but nowhere for the database to store data!
+
+*   Create a volume to persist to for the data:
+
+```shell
+$mac docker volume create content-publisher-postgres
+```
+
+* Restart the postgres container in your other terminal window/tab:
+
+```shell
+docker run -it --rm -v content-publisher-postgres:/var/lib/postgresql/data postgres
+```
+
+But we also need to specify a user to log into the database.
+
+#### Create a database user
+
+*   Add the `builder` user to Dockerfile:
+
+```docker
+RUN useradd -m builder
+USER builder
+```
+
+* Re-build content-publisher container:
+
+```shell
+$mac docker build -t content-publisher
+```
+
+* Re-run docker with a new flag, `--privileged`, which gives the build user elevated permissions to run postgres:
```

Comment:
This is actually to run Chrome, which requires certain low-level syscalls, since we run it in 'sandbox' mode.